### ISSUE TITLE
Support dictionary of nullable items

### ISSUE THREAD

**Issue by matthewz0584** [RicoSuter/NSwag#2116]
I have a model as follows:

```
public class PurchaseReadDto
{
    public Dictionary<string, decimal?> ParticipantsParts { get; set; }
}
```
Swagger for ParticipantsParts of which is generated as:

```
"ParticipantsParts": {
     "type": "object",
     "additionalProperties": {
           "type": "number",
            "format": "decimal"
       }
}

```
And in my C# client it surfaces back as `System.Collections.Generic.IDictionary<string, decimal>`.

As was suggested in https://github.com/RicoSuter/NSwag/issues/82, I can add null to type `("type": ["null", "number"]`) and the generated code turns into `Dictionary<string, decimal?>` - exactly what I need.
But can I make it being generated this way right away?
Maybe the attribute `ItemsCanBeNullAttribute` from https://github.com/RicoSuter/NSwag/issues/1951 could be reused although in this case there is no ambiguity: decimal? must clearly allow null values.

I use this `$(NSwagExe_Core20) webapi2swagger` for swagger generation.

Thanks!

**Comment by RicoSuter** [comment]
You are right, in this case it should just work...

**Comment by matthewz0584** [comment]
That was quick! Thank you very much.

But I can not make it work. I installed the last version of [NJsonSchema](https://github.com/RicoSuter/NJsonSchema/releases/tag/NSwag-Build-1889) package, copied assemblies from there to `nswag.msbuild\12.1.0\tools\NetCore21\` (I dont know any other way to update them for NSwagExe_Core21), got this record in generated swagger `"x-generator": "NSwag v12.1.0.0 (NJsonSchema v9.13.30.0 (Newtonsoft.Json v11.0.0.0))"`, where v9.13.30 is the updated version, but the generated result is still the same. No sign of nullability, with or without ItemsCanBeNullAttribute.

What do I miss?

**Comment by RicoSuter** [comment]
Can you post the spec of this `"ParticipantsParts": {` again? Is it still the same?

**Comment by matthewz0584** [comment]
Absolutely

```
    "PurchaseReadDto2": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "ParticipantsParts": {
          "type": "object",
          "additionalProperties": {
            "type": "number",
            "format": "decimal"
          }
        }
      }
    }
```
and
```
    [System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "9.13.30.0 (Newtonsoft.Json v11.0.0.0)")]
    public partial class PurchaseReadDto2 
    {
        [Newtonsoft.Json.JsonProperty("ParticipantsParts", Required = Newtonsoft.Json.Required.Default, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
        public System.Collections.Generic.IDictionary<string, decimal> ParticipantsParts { get; set; }
        public string ToJson() 
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this);
        }
        public static PurchaseReadDto2 FromJson(string data)
        {
            return Newtonsoft.Json.JsonConvert.DeserializeObject<PurchaseReadDto2>(data);
        }
    }
```
I'll try AspNetCore Middleware (should have done that before) to elliminate possibility of update mess.

**Comment by matthewz0584** [comment]
Same result. I installed NJsonSchema from nuget beside NSwag.AspNetCore
 
![image](https://user-images.githubusercontent.com/2135003/56297316-cbfe1500-6138-11e9-8a06-6f8ae39c0dca.png)

**Comment by RicoSuter** [comment]
Is this Swagger 2 or OpenAPI 3? Because Swagger 2 does not know about null it might not work there...

**Comment by matthewz0584** [comment]
Yes, it was Swagger 2 as it turns out now. I conjectured that if it could generate spec for simple nullable properties it will cope with Dictionary of nullables.

So if I add AddOpenApiDocument to middleware, the generated spec is
```
"PurchaseReadDto2": {
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "participantsParts": {
            "type": "object",
            "nullable": true,
            "additionalProperties": {
              "type": "number",
              "format": "decimal",
              "nullable": true
            }
          }
        }
      }
```
which is probably the right thing.
Can I use this OpenApi in my tool chain the way I used swagger before from MSBuild: webapi2swagger to swagger2csclient? Could you give me any directions?

Thank you for your help!

**Comment by RicoSuter** [comment]
In Swagger 2 we should generate a custom x-nullable property, so this should also work... need to check why this is not the case here.

OpenApi 3 should just work in the toolchain too, but there might be some edge cases which are not covered yet... but it is clearly the stuff which will get most attention in the future. It depends whether other non-nswag tools you’re using already support v3

**Comment by RicoSuter** [comment]
Added another PR, but this might be a major breaking change.

**Comment by matthewz0584** [comment]
Moved to OpenApi and aspnetcore2swagger from swagger and webapi2swagger.

Now I have the spec generated with "nullable" set to true as above, but the generated client code never changed:
```
[System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "9.13.31.0 (Newtonsoft.Json v11.0.0.0)")]
public partial class PurchaseReadDto2 
{
    [Newtonsoft.Json.JsonProperty("ParticipantsParts", Required = Newtonsoft.Json.Required.Default, NullValueHandling = Newtonsoft.Json.NullValueHandling.Ignore)]
    public System.Collections.Generic.IDictionary<string, decimal> ParticipantsParts { get; set; }
```
Is there any setting for this?

**Comment by RicoSuter** [comment]
Wow, you've found extremely nasty but which only applies to additional items and additional properties schemas... will be fixed in nswag v12.2 soon

**Comment by matthewz0584** [comment]
Oh my! I got off easy! It could have attacked me)

All right, I have my workaround ("type": ["null", "number"]) for now.

I am looking forward to when you have fixed it. Many thanks!

**Comment by RicoSuter** [comment]
Your hack is fine for some days, but `type": ["null", "number"]` is not valid in swagger nor in openapi

**Comment by matthewz0584** [comment]
"will be fixed in nswag v12.2 soon" - and it was. Thank you!
